default/customizations/custom_sitreps.txt

```diff
@@ -104,6 +104,45 @@ EffectsGroup
             ]
             empire = Source.Owner
 
+EffectsGroup
+    scope = And [
+        Capital
+        Not OwnedBy empire = Source.Owner
+        Number low = 1 condition = And [
+            // maybe speed up by ORing subcategories
+            OwnedBy empire = Source.Owner
+            VisibleToEmpire empire = RootCandidate.Owner
+        ]
+    ]
+    // Some things need to be fixed: TurnTechResearched should probably not return 0 if not researched
+    // Some things need to be fixed: TurnTechResearched should include tech which is effective this turn
+    activation = Or [
+       [[EMPIRE_TECH_RESEARCHED(SPY_DETECT_2)]]
+       [[EMPIRE_TECH_RESEARCHED(SPY_DETECT_3)]]
+       [[EMPIRE_TECH_RESEARCHED(SPY_DETECT_4)]]
+       [[EMPIRE_TECH_RESEARCHED(SPY_DETECT_5)]]
+    ]
+    // does not help with TurnTechResearch...
+    priority = [[END_CLEANUP_PRIORITY]]
+    effects =
+        GenerateSitRepMessage
+            message = "SITREP_EMPIRE_TECH_RESEARCHED_DETECTION"
+            label = "SITREP_EMPIRE_TECH_RESEARCHED_DETECTION_LABEL"
+            icon = "icons/tech/active_radar.png"
+            parameters = [
+                tag = "empire"  data = Source.Owner
+            ]
+            empire = Target.Owner
+
+EMPIRE_TECH_RESEARCHED
+'''And [
+        (CurrentTurn > 1)
+        Or [
+            (CurrentTurn == TurnTechResearched empire = Source.Owner name = "@1@")
```

Comment:
Why both the turn the tech is researched and the turn after? Does this generate two sitreps saying the same thing, one turn apart?

Reply:
This is a "leftover" from working on shifting the tech researched turn.
The FOCS implementation works for different backend implementations.

With the old and the also new backend implementation both CurrentTurn == TurnTechResearched will never fire.

I left it in intentionally as i think if there are again changes, a double sitrep will more likely be noticed than a missing sitrep so we get a timely bug report in the worst case.

Reply:
Add a comment explaining this then...

Reply:
With the new backend implementation this actually changed. Now i get two sitrep messages.

Currently (11a579785a9e1a067b12d5fb2c75c2cbdcf1608f)  it is like:
If "tech researched" sitrep is shown on turn 5, the "upgraded detection technology" is shown on turn 6, by the  CurrentTurn == TurnTechResearched (plus nothing) condition. The TurnTechResearched for the tech is turn 5. For an initial tech it is -32768 and for an unresearched tech it is zero. 

Besides the break in the continuum (initial tech < no tech < normal tech) this looks quite sensible.

Reply:
I don't understand you point in that comment...

Reply:
It means i need to update this PR because the backend changed and now one gets the sitrep info twice.

Reply:
And it would be good if initial TurnResearched < normal TurnResearched < not researched TurnResearched. 
So for unresearched tech TurnResearched could/should return maximum int. 
So that currentTurn >= TurnResearched means that the tech is actually researched

Reply:
It would be somewhat awkward to add such a special case in the C++ code specifically for `TurnTechResearched`

Reply:
Then we always have to check awkwardly on use site (FOCS)

Reply:
> It would be somewhat awkward to add such a special case in the C++ code specifically for `TurnTechResearched`

I checked the code and now I see what you mean. Because this is generic valueref handling of maps it would be hard to add a value-to-return-if-not-found in a generic and effective way. Probably a separate path in ComplexVariable<int>::Eval  which passes the value-to-return-if-not-found to GetIntEmpirePropertySingleKey.

Reply:
It should be possible to write a macro in FOCS to check if a tech us researched and return a value dependent on that. I think there is an empire has tech researched condition which could be combined with the If ValueRef.

Reply:
Decided to add a special case for TurnTechResearched...

https://github.com/freeorion/freeorion/commit/d4e5b76ff0fe450a7cfa0dad66f454cb9fc81580